syntax = "proto3";

option go_package = "github.com/NethermindEth/juno/internal/services/vmrpc";

message GetValueRequest {
  bytes key = 1;
}

enum NodeType {
    EDGE_NODE = 0;
    BINARY_NODE = 1;
}

message VMTrieNode {
    uint32 len = 1;
    bytes bottom = 2;
    bytes path = 3;
    bytes left = 4;
    bytes right = 5;
    NodeType type = 6;
}

message VMContractState {
    bytes contractHash = 1;
    bytes storageRoot = 2;
    uint32 height = 3;
}

message VMContractDefinition {
    bytes value = 1;
}

service StorageAdapter {
  rpc GetPatriciaNode(GetValueRequest) returns (VMTrieNode) {}
  rpc GetContractState(GetValueRequest) returns (VMContractState) {}
  rpc GetContractDefinition(GetValueRequest) returns (VMContractDefinition) {}
}

message VMCallRequest {
  repeated bytes calldata = 1;
  bytes contract_address = 2;
  bytes class_hash = 3;
  bytes root = 4;
  bytes selector = 5;
  bytes sequencer = 6;
}

message VMCallResponse {
  repeated bytes retdata = 1;
}

service VM {
  rpc Call (VMCallRequest) returns (VMCallResponse) {}
}
